wasm/wasm_source/src/vp_masp.rs

```diff
@@ -55,7 +97,32 @@ fn validate_tx(
         transparent_tx_pool += shielded_tx.value_balance.clone();
 
         // Handle shielding/transparent input
+        // The following boundary conditions must be satisfied
+        // 1. One transparent input
+        // 2. Zero transparent output
+        // 3. the transparent transaction value pool's amount must equal the
+        // containing wrapper transaction's fee amount
         if transfer.source != masp() {
+            // Satisfies 1.
+            if shielded_tx.vin.len() != 1 {
+                debug_log!(
+                    "Transparent input to a transaction from the masp must be \
+                     1 but is {}",
+                    shielded_tx.vin.len()
+                );
+                return reject();
+            }
+
+            // Satisfies 2.
+            if !shielded_tx.vout.is_empty() {
```

Comment:
This check enforces that transactions with a non-MASP source must have no transparent outputs. While this check can be done, the transaction would still be valid (i.e. leaving the shielded pool and transparent pool in a consistent state) without it. Perhaps we can remove this check to simplify the logic?